drltrace/drltrace.cpp

```diff
@@ -79,6 +80,276 @@ static file_t outf;
+}
+
+static int
+split(char *line, const char *delim, std::vector<char *> *tokens_list) {
+    char *token = NULL;
+    int count = 0;
+
+    token = strtok(line, delim);
+    while (token != NULL) {
+        tokens_list->push_back(strdup(token));
+        token = strtok(NULL, delim);
+        count++;
+    }
+    return count;
+}
+
+static bool
+parse_line(char *line, int line_num) {
+    drsys_arg_t *tmp_arg;
+    char *func_name = NULL;
+    int elem_index = 0, tokens_count = 0;
+
+    if (line == NULL)
+        return false;
+
+    if (strlen(line) <= 0 || line[0] == '\n' || line[0] == '\r')
+        return false;
+
+    if (line[0] == ';') /* just a comment */
+        return true;
+
+    std::vector<char*> *tokens = new std::vector<char*>();
+    tokens_count = split(line, ",", tokens);
+
+    if (tokens_count <= 0) {
+        VNOTIFY(1, "unable to parse config file at line %d: %s" NL, line_num, line);
+        delete (tokens);
+        return false;
+    }
+
+    std::vector<drsys_arg_t *> *args_vector = new std::vector<drsys_arg_t *>();
+    std::vector<char*>::iterator it;
+    for (it = tokens->begin(); it != tokens->end(); it++) {
+        /* i#1948: Currently, we don't support ret value printing and
+         * skipping it here. */
+        if (elem_index >= 2) {
+            tmp_arg = config_parse_type(*it, elem_index - 2);
+            args_vector->push_back(tmp_arg);
+        } else if (elem_index == 1)
+            func_name = *it;
+
+        elem_index++;
+    }
+
+    delete tokens;
+
+    if (func_name == NULL || args_vector->size() <= 0) {
+        VNOTIFY(1, "unable to parse config file at line %d: %s" NL, line_num, line);
+        return false;
+    }
+
+    VNOTIFY(2, "adding %s from config file with %d arguments in the hashtable" NL,
+            func_name, args_vector->size());
+    IF_DEBUG(bool ok =)
+        libcalls_hashtable_insert(strdup(func_name), args_vector);
+    ASSERT(ok, "failed to add libcall in the hashtable");
+
+    return true;
+}
+
+static void
+parse_config(void) {
+    uint64 size_to_read = 0;
+    size_t size_read = 0;
+    file_t file_desc = INVALID_FILE;
+    int lines_count = 0, line_num = 1;
+    char *buf = NULL;
+
+    if (op_disable_config.get_value() == true)
+        return;
+
+    /* open and read config file */
+    file_desc = dr_open_file(op_config_file.get_value().c_str(), DR_FILE_READ);
+    if (file_desc == INVALID_FILE) {
+        VNOTIFY(0, "unable to open config file at %s, config is not used" NL,
+                    op_config_file.get_value().c_str());
+        op_disable_config.set_value(true);
+        return;
+    }
+
+    if (dr_file_size(file_desc, &size_to_read) == false) {
+        VNOTIFY(0, "unable to open config file at %s, config is not used" NL,
+                    op_config_file.get_value().c_str());
+        op_disable_config.set_value(true);
+        dr_close_file(file_desc);
+        return;
+    }
+
+    buf = (char *)global_alloc(size_to_read, HEAPSTAT_MISC);
```

Comment:
Generally our approach to parsing text files in existing code (drsymcache, Dr. Memory suppression files, drcovlib, etc.) is to mmap the file and then parse it as a const string